src/libraries/System.Net.Http/tests/FunctionalTests/Socks/SocksProxyTest.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Test.Common;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Http.Functional.Tests.Socks
+{
+    public abstract class SocksProxyTest : HttpClientHandlerTestBase
+    {
+        public SocksProxyTest(ITestOutputHelper helper) : base(helper)
+        {
+        }
+
+        [Theory]
+        [InlineData("socks4")]
+        [InlineData("socks4a")]
+        [InlineData("socks5")]
+        public async Task TestLoopbackAsync(string schema)
+        {
+            await LoopbackServerFactory.CreateClientAndServerAsync(async url =>
+            {
+                using (var proxy = LoopbackSocksServer.Create())
+                {
+                    using (var handler = CreateHttpClientHandler())
+                    using (var client = CreateHttpClient())
+                    {
+                        client.DefaultRequestVersion = UseVersion;
+                        handler.Proxy = new WebProxy($"{schema}://localhost:{proxy.Port}");
+
+                        string response = await client.GetStringAsync(url);
+                        Assert.Equal("Echo", response);
+                    }
+                }
+            },
+            async server => await server.HandleRequestAsync(content: "Echo"));
+        }
+
+        [Theory]
+        [InlineData("socks4")]
+        [InlineData("socks4a")]
+        [InlineData("socks5")]
+        public async Task TestLoopbackHttpsAsync(string schema)
+        {
+            using var cert = TestHelper.CreateServerSelfSignedCertificate();
+
+            await LoopbackServerFactory.CreateClientAndServerAsync(async url =>
+            {
+                using (var proxy = LoopbackSocksServer.Create())
+                {
+                    using (var handler = CreateHttpClientHandler())
+                    using (var client = CreateHttpClient())
+                    {
+                        client.DefaultRequestVersion = UseVersion;
+                        handler.Proxy = new WebProxy($"{schema}://localhost:{proxy.Port}");
+                        handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
```

Comment:
Why is this failing with certificate error?

Reply:
You should pass the handler to `CreateHttpClient` - right now the empty overload is creating its own handler, so you are setting properties on a handler that isn't actually being used.

Reply:
@MihaZupan I need some help debugging the test. It fails for some configurations but success for some. It passes in local VS debug so I have no idea how to debug it.

Reply:
The failures are limited to `socks4a`. I will look into it more

Reply:
The reason the tests were failing is because the loopback server used `AsSpan(length)` instead of `AsSpan(0, length)` when reading the domain name for Socks4a.

A few other details:
- `UseSsl` isn't `false` by default, so `TestLoopbackAsync` wasn't testing the `http` scenario
- When setting it to `false`, you have to set the `HttpVersionPolicy.RequestVersionExact` for H2C

I pushed the test/loopback server change to the PR branch

Reply:
Hey @huoyaoyuan, do you need any more help here to move the PR forward?

Reply:
Yes. I have no idea about the remaining test failures.